### PR TITLE
{discovery,remote_write}/azure: Support default SDK authentication

### PR DESCRIPTION
**Pull request by jkroepke**
This PR expose the [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#defaultazurecredential) chain offered by azidentity packag which is introduced in #11860 

It support authentication via environment variables and support following auth schemas:

* Environment - DefaultAzureCredential will read account information specified via [environment variables](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#environment-variables) and use it to authenticate.
* Workload Identity - If the app is deployed on Kubernetes with environment variables set by the workload identity webhook, DefaultAzureCredential will authenticate the configured identity.
* Managed Identity - If the app is deployed to an Azure host with managed identity enabled, DefaultAzureCredential will authenticate with it.
* Azure CLI - If a user or service principal has authenticated via the Azure CLI az login command, DefaultAzureCredential will authenticate that identity.

DefaultAzureCredential work like a charm together with [Azure Workload Identity Webhook](https://github.com/Azure/azure-workload-identity).

### CODE REVIEW

**Comment by jkroepke** [comment]
I test it locally, using

```
scrape_configs:
  - job_name: azure
    azure_sd_configs:
      - authentication_method: SDK
        subscription_id: "1988b893-553c-4652-bd9b-52f089b21ead"

```

and I was able to authenticate with SDK default token chain. In my case, Azure CLI was used to authenticate.

**Comment by jkroepke** [comment]
cc @mmorel-35 since you do #11860

**Review by mmorel-35 — APPROVED** [review]
LGTM

**Comment by jkroepke** [comment]
```
--- FAIL: TestTokenProvider (12.02s)
    --- FAIL: TestTokenProvider/TestNewTokenProvider (12.01s)
        azuread_test.go:316: 
            	Error Trace:	/__w/prometheus/prometheus/storage/remote/azuread/azuread_test.go:316
            	Error:      	Expected nil, but got: &errors.errorString{s:"Failed to get access token: access token expiry is less than the current time"}
            	Test:       	TestTokenProvider/TestNewTokenProvider
FAIL
FAIL	github.com/prometheus/prometheus/storage/remote/azuread	12.047s
```

I may need assistance here.

**Comment by mmorel-35** [comment]
Have you tried to inline testTokenExpiry value ?
Having a var at this level might cause to create a flaky token

**Comment by jkroepke** [comment]
Yes, it has no affect.

**Comment by jkroepke** [comment]
The error is not related to the SDK auth. if I have 2 provider configured in the test loop, its fine. Adding third one results into this.

**Comment by roidelapluie** [comment]
Could we try:

-var testTokenExpiry = time.Now().Add(5 * time.Second)                                                                                                
+func testTokenExpiry() time.Time { return time.Now().Add(5 * time.Second) }

**Comment by jkroepke** [comment]
@roidelapluie no difference.

**Comment by jkroepke** [comment]
Added `.Once()` to mockCredential call solves the issue

**Comment by rakshith210** [comment]
Thanks for the PR @jkroepke. It looks good overall. There are couple of things I wanted to share especially in regard to the remote write section:
1. There is already credential client present for managed identity and client secret. The reason for going this route over the default credential was that it would be intuitive for the user, especially ones who are new to Azure and Azure auth, by providing explicit config block: `managed_identity` and `oauth`, where they could provide the values required for each authentication type. This ensures that the user is deliberate about which auth type they are using and not worry about any additional environment variables they have to set or have any webhook installed and have all configuration required to work in one place i.e., Prometheus configuration. Based on this, my suggestion would be to creating a separate workload identity credential to add workload identity support in a similar way as I mentioned over here: #13061
2.   DefaultCredential too is helpful in certain scenarios and would be a good to have feature. Another suggestion here would be to call this method of authentication as `default` or `default_auth` rather than `sdk` as other existing credentials too are from the same azidentity package and better helps with distinction from other auth types.

**Comment by jkroepke** [comment]
Thanks for sharing m.

1. I guess an other point is that the old Azure SDK doesnt provide such an default credential. On top, the current design require to configure secrets inside the configuration which flag the while config as secret. Its also discussable what is intuitive. Almost each application supports the AZURE_ environment variables, which is perfect for Kubernetes. The mentioned Webhook is preinstalled on Azure Kubernetes Service and the [documentation](https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview) around Workload Identity depends on a Webhook. Please mention that Token Volume is required in addition to environment variables which increase the complexlity a lot. 
I still see some use-case described in #13061 and the proposal looks fine, but that an other Pull Request. The scope here of the PR is to integrate the DefaultTokenCredential of Azure SDK, not a generic interface for Workload Identity.

2. From user point of view, it heavly confuses me, if there in an option called default which is not the default option in Prometheus. If there is an oppunity to change the default option from oauth to this, it would be heavly increase the out of the box experience for azure users which can already enjoy from a lot of other applications.

I got the naming from Grafana. The AWS datasource has such an name for SDK native auth. But I'm open for any name. I don't have strong feelings.

**Comment by roidelapluie** [comment]
I would tend to merge this pull request now. I think the current PR is good and I like "SDK"

**Comment by jkroepke** [comment]
@roidelapluie For transparency, I could only test discovery using SDK auth. And It worked fine in my case. I don't have an Prometheus with requires Azure Auth.

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -3565,6 +3567,11 @@ azuread:
       [ client_secret: <string> ]
       [ tenant_id: <string> ] ]
 
+  # Azure SDK auth.
+  # See https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication
+  [ sdk:
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:3572]: You need to indent this

Comment (jkroepke) [reply docs/configuration/configuration.md:3572]: looks fine to me

**Comment by rakshith210** [comment]
@jkroepke I wanted to clarify few things in regard to your comment earlier.
1. `NewDefaultAzureCredential` was available since the inception of the azidentity library back in 2020 and not something that came with the new version of the package. Also, workload identity webhook is not pre-installed with Azure Kubernetes Service, but something the user would need to install before using workload identity along with enabling OIDC issuer in their cluster. I want to clarify here that Azure workload identity is not just applicable to Azure Kubernetes service but can work with any Kubernetes cluster on-prem or other clouds so we would need to be mindful of that as well. The new feature `NewDefaultAzureCredential` predominately brings in remote write is the workload identity support as managed identity and client secret(oauth) is already present hence, my previous comment centered around workload identity support. If we do introduce `NewDefaultAzureCredential` and later introduce `WorkloadIdentityCredential`, we would be exposing two different experiences to Prometheus users using Azure auth; `WorkloadIdentitycredential`, `ManagedIdentityCredential`(already exists) and `ClientSecretCredential`(already exists) would have all configuration part of Prometheus remote write configuration whereas `DefaultAzureCredential` would have configuration in Prometheus remote write as well as in deployment yaml of the Prometheus container where they would need to add environment variables required. I just want to iterate again that the changes you have for `DefaultAzureCredential` looks good but I wanted to highlight the nuances the Prometheus users are going to experience with Azure auth and avoid any confusion due to introduction of new pattern for existing users.
2. Yea I agree `default` sounds too generic and might lead to confusion. My intention basically was to convey user what kind of auth they are opting in this case and I felt `sdk` to be very abstract unless someone knows about the inner workings. We could possibly consider `chainedauth`, just a suggestion ?

@LeviHarrison / @csmarchbanks - I would love to get your thoughts too on this.

**Comment by jkroepke** [comment]
I'm not against declarative configuration for Workload Identity. It from my point of view a separate PR.

For the integration of proposal, the test scope would be heavly increased. I'm able to cover Azure here, but not other Integration.

Also mention that that this PR brings authentication via local azure cli which is helpful for local development.

The Grafana eco system supports both. Explicit configuration or via environment.

Another PR at Thanos follows the same schema + change the auth default to AzureDefaultCredential (https://github.com/thanos-io/objstore/pull/82). A lot of tooling around the Prometheus eco system supporting environment variables and from point of view a straigt forward, future proof and maintain-less option.

It would great, if Prometheus would offer the same. Explicit configuration can be provided by an separate PR.

**Comment by jkroepke** [comment]
Hi all, how I can help here to move forward?

**Comment by jkroepke** [comment]
@roidelapluie is there anything what I can do here to more forward?

**Comment by amirschw** [comment]
Looking forward for this pull request to be completed so we can use Prometheus with Workload Identity in our AKS clusters. @roidelapluie your assistance will be appreciated 🙏🏻

**Review by bboreham — APPROVED** [review]
Code looks good to me; we don't have anyone in Prometheus team who knows Azure so approving on limited information.

**Pull request merged**